### DISCUSSION TITLE
How to consume ODataV4 web Api

### DISCUSSION THREAD

**Discussion by e066377** [unoplatform/uno#6836] [Q&A]
## What would you like clarification on:
Hello,
I want to use Microsoft.OData.Client for UWP and WASM platforms to generate models from ODatV4 web Api with INotifyPropertyChanged interface implemented. It works without problem for UWP platfrom but for WASM it is not very clear how to introduce httpclient to the container. Setting 
 Default.Container HttpRequestTransportMode = HttpRequestTransportMode.HttpClient is not enough;

For LOB applications ODatav4 web Api is more powerful than ordinary web Api, so it will be very helpful if you can add an how-to tutorial for ODatav4 web Api usage.

## Concern?
- [X] Usage in industry
- [x] Clarification of capabilities
- [X] Getting started with Uno
- [X] Developing with Uno

## For which Platform:
- [X] WebAssembly
- [X] Windows

**Comment by jeromelaban** [comment]
Thanks for the report. 

The best way for us to help you there is to create a sample app with the Uno template, and make it work for UWP. From that point, we'll look at the working solution and try to apply it to Wasm.

**Comment by e066377** [comment]
[OData.Uno.zip](https://github.com/unoplatform/uno/files/7036936/OData.Uno.zip)

Hi.
I have found a solution though not sure if it is the best way, Simple.OData.Client.
Simple.OData.Client ODataClientSettings has a property HttpClient which can be set by property assignment or constructor argument and it works. The only downside is you have to implement INotifyPropertyChanged yourself or use Simple.OData.Client and Microsoft.OData.Client together.

The attached sample project uses Simple.OData.Client and Microsoft.OData.Client together.
For Simple.OData.Client both UWP and WASM can call OData api and load data, but for  Microsoft.OData.Client it is not clear how to use custom HttpClient. Custom HttpClient is needed not only to load data but also to set auth headers for a real world application.

In constructing this application,  api metada (https://localhost:44352/odata/$metadata) is saved as an xml file named as "metadata.xml"  in Sever application DataSource folder and OData Connected service is added to UWP app, setting address to this file in configuration which should be configured for an updated xml file or new project.  Then the generated OData Service reference folder is moved to Shared project and removed form UWP.

To make Simple.OData.Client work Simple.OData.Client.V4.Adapter assemly is added to LinkerConfig.xml file and a static method is added in WASM.Program class that calls Simple.OData.Client.V4Adapter.Reference() method, else Simple.OData.Client.V4Adapter is not loaded. https://github.com/simple-odata-client/Simple.OData.Client/issues/53 describes reason.

Regards.
Ali Arslan
